[CORE-13247] cloud_topics: add initial L0 GC framework

dotnwat opened the pull request:
Introduces the core L0 GC framework. Subsequent work is required to complete L0 GC but are out of scope of this PR.

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [x] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v25.2.x
- [ ] v25.1.x
- [ ] v24.3.x

## Release Notes

<!--
If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

Otherwise, adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

### Bug Fixes

* Short description of the bug fix if this is a PR to `dev` branch.

### Features

* Short description of the feature. Explain how to configure.

### Improvements

* Short description of how this PR improves existing behavior.

-->

* none

dotnwat committed:
cloud_topics: add level zero gc worker framework

Adds the core worker framework for level zero garbage collection. The
worker is responsible for coordinating and driving the gc process by
querying for the same to delete epoch and polling the configured object
storage system for new data eligible for collection.

External functionality is abstracted away behind provider interfaces.
This commit only adds implementations for these interfaces sufficient
for testing.

Signed-off-by: Noah Watkins <noahwatkins@gmail.com>

dotnwat committed:
cloud_topics: add initial L0 gc provider implementations

Adds an initial and incomplete implementation of real providers required
by the level zero gc process. The primary missing component is the full
calculation of the safe-to-delete epoch. Additional work is required to
support this.

Signed-off-by: Noah Watkins <noahwatkins@gmail.com>

dotnwat committed:
cloud_topics: add manager

Introduces a top-level manager component for cloud topics. The idea is
for the manager to coordinate work and manage state that is inherently
global to cloud topics. The current user is the level zero garbage
collection process that need only have a location in the cluster to run,
so it attaches to the active manager instance.

The manager itself as well as the level zero GC process are stateless.
Because of the stateless nature, an entirely new internal, dedicated
topic/partition is avoided and instead the manager attaches to
partition-0 of the L1 domains topic. Once the manager or its dependents
become stateful or for whatever reason we want to avoid attaching to the
L1 domains NTP, then the attach target NTP in the mananger can be updated.

For now using the L1 domains topic coordinates nicely with the
dependencies of L0 GC process because this topic must exist before any
L0 data ever becomes eligible for GC.

Signed-off-by: Noah Watkins <noahwatkins@gmail.com>

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--a0c68c75-7852-4c53-9d80-f92d556bb29e-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/71540">build#71540</a></summary>


| test_class                              | test_method                | test_arguments   | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                     |
|:----------------------------------------|:---------------------------|:-----------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------|
| InternalTopicProtectionLargeClusterTest | test_consumer_offset_topic | null             | integration | https://buildkite.com/redpanda/redpanda/builds/71540#0198f211-b911-4ee2-b75f-59c1212870c0 | FLAKY         | 19/21    | upstream reliability is '100.0'. current run reliability is '90.47619047619048'. drift is 9.52381 and the allowed drift is set to 50. The test should PASS |
</details>
<!--/body-->

rockwotj commented on src/v/cloud_topics/level_zero/gc/tests/level_zero_gc_tests.cc:
```diff
@@ -0,0 +1,178 @@
+    cloud_topics::level_zero_gc gc;
+};
+
+template<typename Func>
+::testing::AssertionResult Eventually(Func func) {
+    int retries = 50;
+    while (retries-- > 0) {
+        if (func()) {
+            return ::testing::AssertionSuccess();
+        }
+        seastar::sleep_abortable(std::chrono::milliseconds(100)).get();
+    }
+    return ::testing::AssertionFailure() << "Timeout";
+}
```
Ohhh - very nice. I like this more than the macro we have

rockwotj reviewed (dismissed):
Very nice. LGTM. No comments!

andrwng commented on src/v/cloud_topics/level_zero/gc/level_zero_gc.cc:
```diff
@@ -0,0 +1,231 @@
+        const auto object_epoch = object_path_factory::level_zero_path_to_epoch(
+          object.key);
+
+        // validate expected L0 object name format, and extract epoch
+        if (!object_epoch.has_value()) {
+            vlog(
+              cd_log.error,
+              "Unable to parse epoch during L0 GC: {}",
+              object_epoch.error());
+            co_return std::unexpected(collection_error::invalid_object_name);
```
Stopping all progress here seems like an explicit decision. Is it required for correctness? Do we break any invariants if we continue?

dotnwat replied:
It's very conservative, yeh. The idea here is basically that if we ask S3 to return only objects with the prefix `/path/to/L0/data/objects/...` and it returns object names whose formats are unexpected (for that prefix), what should we do? You're right we can skip them. But why are they there? Are we seeing the result of some bug elsewhere?

andrwng replied:
Yea that's fair. I just had the worry that if I wanted to I could bork GC by manually dropping in an object.

Though maybe that's also a feature. In any case, I don't think it' something we need to change right now, but it is something to consider.

dotnwat replied:
>manually dropping in an object.

yes, that's true. i think it being conservative is a feature at this early stage. one case where i could see us wanting a relaxed behavior (in addition to the case you mentioned) is in scenarios where we are changing the naming format and the old format is still valid.

andrwng commented on src/v/cloud_topics/level_zero/gc/level_zero_gc.cc:
```diff
@@ -0,0 +1,231 @@
+            co_return std::unexpected(collection_error::invalid_object_name);
+        }
+
+        // detect non-lexiographic ordering. this may indicate that GC will not
```
nit: "lexio" typo in a few places

andrwng commented on src/v/cloud_topics/level_zero/gc/level_zero_gc.cc:
```diff
@@ -0,0 +1,231 @@
+            static seastar::logger::rate_limit rate(rate_limit);
+            vloglr(
+              cd_log,
+              seastar::log_level::warn,
```
Should this be an error so we can detect it in CI?

I guess it doesn't necessarily hurt correctness, but I can imagine in the worst case of non-lexicographic ordering, this loop could get stuck iterating over new objects, and never make progress onto the oldest objects

dotnwat replied:
Yeh I automatically chose `warn` because it wasn't a fatal case, but you're right, we really do want to know the moment this is violated because it is a signal that we need to introduce a more general form of GC progress tracking.

andrwng commented on src/v/cloud_topics/manager/manager.cc:
```diff
@@ -0,0 +1,120 @@
+  , group_manager_(group_manager)
+  , level_zero_gc_(remote_, bucket_) {}
+
+seastar::future<> cloud_topics_manager::start() {
```
Just making sure I understand, with this change, could we add a ducktape test for GC? Or is anything else missing?

dotnwat replied:
Not quite yet.

1. The safe-to-delete epoch is pinned in this PR at `-1`. There are some supportive APIs in the controller that we need to add to have a real epoch value.
2. We don't have the reconciler yet--we need to actually advance the max collectible (or whatever we are calling it nowadays) before there is anything that can be GCd.

andrwng reviewed (commented):
Nice, nothing too major from me, just some questions

dotnwat committed:
cloud_topics: fix typos

Signed-off-by: Noah Watkins <noahwatkins@gmail.com>

dotnwat committed:
cloud_topics: log as error on apparent non-lexicographic ordering

We really want to know if this is happens so we can set out to create a
more general form of GC progress tracking.

Signed-off-by: Noah Watkins <noahwatkins@gmail.com>
